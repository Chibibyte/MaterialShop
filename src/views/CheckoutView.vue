<template>
  <div>
    <CheckoutNavbar />
    <v-main>
      <router-view />
      <!-- LEAVE MODAL -->
      <v-dialog v-model="leaving" max-width="290">
        <v-card>
          <v-card-title class="headline">Leave checkout?</v-card-title>
          <v-card-text>Your checkout data will be lost.</v-card-text>
          <v-card-actions>
            <v-spacer></v-spacer>
            <v-btn color="green darken-1" text @click="leaving = false"
              >Cancel</v-btn
            >
            <v-btn color="green darken-1" text @click="leave">Confirm</v-btn>
          </v-card-actions>
        </v-card>
      </v-dialog>
    </v-main>
  </div>
</template>
<script>
/**
 * Business Logic
 * -display CheckoutNavbar
 * -contain router-view for CheckoutViews
 *
 * Inputs:
 *    -$store.state.checkout.leaving
 * Outputs:
 *    -leaving
 */
import CheckoutNavbar from "@/components/CheckoutNavbar.vue";
export default {
  name: "CheckoutView",
  data() {
    return {};
  },
  components: {
    CheckoutNavbar,
  },
  computed: {
    leaving: {
      get() {
        return this.$store.state.checkout.leaving;
      },
      set(value) {
        this.$store.commit("checkout/setLeaving", value);
      },
    },
  },
  methods: {
    leave() {
      this.$store.commit("checkout/clearCheckoutContent");
      this.$store.commit("checkout/clearCheckoutStates");
      this.$router.push("/");
      this.leaving = false;
    },
  },
  beforeMount() {
    if (this.$store.state.noncheckout.cart.length === 0) {
      this.$router.push("/");
    }
  },
};
</script>