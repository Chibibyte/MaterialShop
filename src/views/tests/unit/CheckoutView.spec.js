import Vuex from 'vuex'
import { shallowMount, createLocalVue } from '@vue/test-utils';
import CheckoutView from '@/views/CheckoutView.vue';

const localVue = createLocalVue();
localVue.use(Vuex);

let store;
let state;
let modules;
let mutations;
let mocks;
let stubs = [
    'router-view',
    'v-main',
    'v-dialog',
    'v-card',
    'v-card-title',
    'v-card-text',
    'v-card-actions',
    'v-spacer',
    'v-btn'
]
let wrapper;

beforeEach(() => {
    mutations = {
    }

    state = {
    }

    modules = {
        checkout: {
            namespaced: true,
            state() {
                return {
                    leaving: false
                }
            },
            mutations: {
                setLeaving: jest.fn(),
                clearCheckoutContent: jest.fn(),
                clearCheckoutStates: jest.fn()
            }
        },
        noncheckout: {
            namespaced: true,
            state() {
                return {
                    cart: []
                }
            }
        }
    }

    mocks = {
        $router: { push: jest.fn() }
    };
    store = new Vuex.Store({
        mutations,
        state,
        modules
    })
    wrapper = shallowMount(CheckoutView, { mocks, stubs, store, localVue });
})

describe('Component', () => {
    test('is instantiated', () => {
        expect(wrapper.isVueInstance).toBeTruthy();
    });

    test('default snapshot', () => {
        expect(wrapper.element).toMatchSnapshot();
    })
});


describe('mounted: ', () => {
    test('CheckoutNavbar', () => {
        expect(wrapper.findComponent({ name: 'CheckoutNavbar' }).exists()).toBe(true)
    })

    test('function: leaving: initial call', () => {
        expect(wrapper.vm.leaving).toBeFalsy();
    })

    test('default', () => {
        expect(mocks.$router.push.mock.calls).toHaveLength(1);
    })
})

describe('getter: leaving', () => {
    test('default', () => {
        expect(wrapper.vm.leaving).toBeFalsy();
    })

    test('leaving in store is true', () => {
        state.checkout.leaving = true;
        expect(wrapper.vm.leaving).toBeTruthy();
        expect(wrapper.element).toMatchSnapshot();
    })
})

describe('setter: leaving', () => {
    test('set leaving true', () => {
        wrapper.vm.leaving = true;
        expect(modules.checkout.mutations.setLeaving.mock.calls).toHaveLength(1);
        expect(modules.checkout.mutations.setLeaving.mock.calls[0][1]).toBeTruthy();
    })

    test('set leaving false', () => {
        wrapper.vm.leaving = false;
        expect(modules.checkout.mutations.setLeaving.mock.calls).toHaveLength(1);
        expect(modules.checkout.mutations.setLeaving.mock.calls[0][1]).toBeFalsy();
    })
})

describe('method: leave', () => {
    test('default', () => {
        mocks.$router.push.mockClear()
        wrapper.vm.leave();
        expect(modules.checkout.mutations.clearCheckoutContent.mock.calls).toHaveLength(1);
        expect(modules.checkout.mutations.clearCheckoutStates.mock.calls).toHaveLength(1);
        expect(mocks.$router.push.mock.calls).toHaveLength(1);
    })
})