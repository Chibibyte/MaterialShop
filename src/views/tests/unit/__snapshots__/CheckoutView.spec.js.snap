// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component default snapshot 1`] = `
<div>
  <checkoutnavbar-stub />
   
  <v-main-stub>
    <router-view-stub />
     
    <v-dialog-stub
      max-width="290"
    >
      <v-card-stub>
        <v-card-title-stub
          class="headline"
        >
          Leave checkout?
        </v-card-title-stub>
         
        <v-card-text-stub>
          Your checkout data will be lost.
        </v-card-text-stub>
         
        <v-card-actions-stub>
          <v-spacer-stub />
           
          <v-btn-stub
            color="green darken-1"
            text=""
          >
            Cancel
          </v-btn-stub>
           
          <v-btn-stub
            color="green darken-1"
            text=""
          >
            Confirm
          </v-btn-stub>
        </v-card-actions-stub>
      </v-card-stub>
    </v-dialog-stub>
  </v-main-stub>
</div>
`;

exports[`getter: leaving leaving in store is true 1`] = `
<div>
  <checkoutnavbar-stub />
   
  <v-main-stub>
    <router-view-stub />
     
    <v-dialog-stub
      max-width="290"
    >
      <v-card-stub>
        <v-card-title-stub
          class="headline"
        >
          Leave checkout?
        </v-card-title-stub>
         
        <v-card-text-stub>
          Your checkout data will be lost.
        </v-card-text-stub>
         
        <v-card-actions-stub>
          <v-spacer-stub />
           
          <v-btn-stub
            color="green darken-1"
            text=""
          >
            Cancel
          </v-btn-stub>
           
          <v-btn-stub
            color="green darken-1"
            text=""
          >
            Confirm
          </v-btn-stub>
        </v-card-actions-stub>
      </v-card-stub>
    </v-dialog-stub>
  </v-main-stub>
</div>
`;
